### PR TITLE
[llvm] Move intrinsics definitions into a template header file llvm-intrinsics.h.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
Makefile.am.in    |    1 
 llvm-intrinsics.h |  174 ++++++++
 mini-llvm-cpp.cpp |   99 ++++
 mini-llvm-cpp.h   |   13 
 mini-llvm.c       | 1146 +++++++++++++-----------------------------------------
 5 files changed, 569 insertions(+), 864 deletions(-)

File: mono/mini/mini-llvm-cpp.cpp [hunk 1]
```diff
@@ -602,3 +602,102 @@ mono_llvm_check_cpu_features (const CpuFeatureAliasFlag *features, int length)
...
+		f = Intrinsic::getDeclaration (unwrap (module), intrins_id, arr);
+		break;
+	}
+	default:
+		g_assert_not_reached ();
+		break;
+	}
+
+	return wrap (f);
+}
```
Comment (imhameed) [inline mono/mini/mini-llvm-cpp.cpp]:
```c++
LLVMValueRef
mono_llvm_register_overloaded_intrinsic (LLVMModuleRef module, IntrinsicId id, LLVMTypeRef *types, int ntypes)
{
    const int max_types = 5;
    if (ntypes > max_types) g_assert_not_reached();
    Type *arr[max_types];
    for (int i = 0; i < ntypes; ++i) arr [i] = unwrap (types [i]);
    auto f = Intrinsic::getDeclaration (unwrap (module), intrins_id, { arr, ntypes });
    return wrap (f);
}
```
??

File: mono/mini/mini-llvm.c [hunk 2]
```diff
@@ -9293,271 +9157,75 @@ mono_llvm_emit_call (MonoCompile *cfg, MonoCallInst *call)
...
-		AddFunc (module, name, LLVMDoubleType (), params, 3);
-		break;
-	}
-	case INTRINS_FMAF: {
-		LLVMTypeRef params [] = { LLVMFloatType (), LLVMFloatType (), LLVMFloatType () };
-
-		AddFunc (module, name, LLVMFloatType (), params, 3);
-		break;
-	}
+	case INTRINS_FMA:
```
Comment (imhameed) [inline mono/mini/mini-llvm.c]: Wouldn't `INTRINS_FMA` need `add_intrins3 (module, id, LLVMDoubleType (), LLVMDoubleType (), LLVMDoubleType ())`?

File: mono/mini/mini-llvm.c [hunk 3]
```diff
@@ -9568,10 +9236,12 @@ add_intrinsic (LLVMModuleRef module, int id)
...
 	case INTRINS_CEIL:
-	case INTRINS_FABS: {
-		AddFunc1 (module, name, LLVMDoubleType (), LLVMDoubleType ());
+	case INTRINS_FABS:
+	case INTRINS_COPYSIGN:
+	case INTRINS_POW:
+		intrins = add_intrins1 (module, id, LLVMDoubleType ());
 		break;
-	}
+	case INTRINS_FMAF:
```
Comment (imhameed) [inline mono/mini/mini-llvm.c]: Same as `INTRINS_FMA`, but with `LLVMFloatType ()` instead.

File: mono/mini/mini-llvm.c (hunk 2 above)
Comment (vargaz) [reply mono/mini/mini-llvm.c]: It only needs the extra parameters, not the whole signature, i.e. in llvm its named
llvm.fma.f64.

Comment (imhameed) [reply mono/mini/mini-llvm.c]: Ah, got it.

(https://github.com/dotnet/llvm-project/blob/release/6.x/llvm/lib/IR/Function.cpp#L597)

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**